packages/components/pages/f-contact-preferences/src/components/ContactPreferences.vue

```diff
@@ -110,50 +121,77 @@ export default {
 
     data () {
         return {
+            isFormDirty: false,
             error: {},
-            preferences: [],
             showErrorPage: false,
-            tenantConfigs
+            tenantConfigs,
+            contactPreferencesApi: new ContactPreferencesApi({
+                httpClient: this.$http,
+                cookies: this.$cookies,
+                baseUrl: this.smartGatewayBaseUrl
+            }),
+            isFormSubmitting: false
         };
     },
 
-    mounted () {
-        try {
-            const data = {
-                preferencesVersionViewed: 0,
-                preferences: [
-                    {
-                        displayName: 'orderUpdates',
-                        email: true,
-                        key: 'orderUpdates',
-                        push: false,
-                        sms: true,
-                        sort: 0
-                    },
-                    {
-                        displayName: 'newsletter',
-                        email: false,
-                        key: 'newsletter',
-                        push: false,
-                        sms: false,
-                        sort: 0
-                    }
-                ]
-            };
-
-            this.preferences = mapToPreferencesViewModel(data).preferences;
-        } catch (error) {
-            this.handleErrorState(new GetPreferencesError(error.message, error?.response?.status));
-        }
+    computed: {
+        ...mapState('fContactPreferencesModule', [
+            'preferences'
+        ])
+    },
+
+    async mounted () {
+        await this.initialise();
     },
 
     methods: {
+        ...mapActions('fContactPreferencesModule', [
+            'loadPreferences',
+            'savePreferences',
+            'editPreference'
+        ]),
+
         handleErrorState (error) {
             this.showErrorPage = true;
             this.error = error;
         },
 
-        onFormSubmit () {}
+        editPreferenceValue (key, field, value) {
+            this.editPreference({ key, field, value });
+            this.isFormDirty = true;
+        },
+
+        setSubmittingState (isFormSubmitting) {
+            this.isFormSubmitting = isFormSubmitting;
+        },
+
+        async initialise () {
+            try {
+                await this.loadPreferences({ api: this.contactPreferencesApi, authToken: this.authToken });
+                this.isFormDirty = false;
+            } catch (error) {
+                this.handleErrorState(new GetPreferencesError(error.message, error?.response?.status));
+            } finally {
+                this.$parent.$emit(STOP_LOADING_SPINNER_EVENT);
+            }
+        },
+
+        async onFormSubmit () {
+            if (!this.isFormDirty) {
```

Comment:
This is nice and defensive, but from a UX perspective what do you think about also disabling the submit button until the form is dirtied?

Reply:
It wasn't a requirement to disable the submit button and causes very little overhead when clicked.  If I do wire it up to the _state_ of the form then it increases the amount of testing (inc. visual) we need to do so to keep things simple I left as is atm 👍